flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java

```diff
@@ -83,27 +86,37 @@ default void validate(TableSchema schema) throws ValidationException {}
     }
 
     /**
-     * Quotes the identifier. This is used to put quotes around the identifier in case the column
-     * name is a reserved keyword, or in case it contains characters that require quotes (e.g.
-     * space). Default using double quotes {@code "} to quote.
+     * Quotes the identifier.
+     *
+     * <p>Used to put quotes around the identifier if the column name is a reserved keyword or
+     * contains characters requiring quotes (e.g., space). Default using double quotes {@code "} to
+     * quote.
+     *
+     * @return the quoted identifier.
      */
     default String quoteIdentifier(String identifier) {
         return "\"" + identifier + "\"";
     }
 
     /**
-     * Get dialect upsert statement, the database has its own upsert syntax, such as Mysql using
-     * DUPLICATE KEY UPDATE, and PostgreSQL using ON CONFLICT... DO UPDATE SET..
+     * Constructs the dialects upsert statement if supported; such as MySQL's {@code DUPLICATE KEY
+     * UPDATE}, or PostgreSQLs {@code ON CONFLICT... DO UPDATE SET..}.
+     *
+     * <p>If the dialect does not support native upsert statements, the writer will fallback to
+     * {@code SELECT} + {@code Update}/{@code INSERT} which may have poor performance.
      *
-     * @return None if dialect does not support upsert statement, the writer will degrade to the use
-     *     of select + update/insert, this performance is poor.
+     * @return The upsert statement if supported, otherwise None.
      */
     default Optional<String> getUpsertStatement(
             String tableName, String[] fieldNames, String[] uniqueKeyFields) {
```

Comment:
Given the performance impact if this is left unimplemented, is it maybe worth not default-implementing it to force implementations to make a conscious decision?

Reply:
agreed